apisix/plugins/body-transformer.lua

```diff
@@ -0,0 +1,206 @@
+            end
+        end
+    end
+    return tbl
+end
+
+
+local decoders = {
+    xml = function(data)
+        local handler = xmlhandler:new()
+        local parser = xml2lua.parser(handler)
+        local ok, err = pcall(parser.parse, parser, data)
+        if ok then
+            return remove_namespace(handler.root)
+        else
+            return nil, err
+        end
+    end,
+    json = function(data)
+        return core.json.decode(data)
+    end,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function transform(conf, body, typ, ctx)
+    local out = {_body = body}
+    if body then
+        local err
+        local format = conf[typ].input_format
+        if format then
+            out, err = decoders[format](body)
+            if not out then
+                err = str_format("%s body decode: %s", typ, err)
+                core.log.error(err, ", body=", body)
+                return nil, 400, err
+            end
+        end
+    end
+
+    local text = conf[typ].template
+    text = decode_base64(text) or text
+    local ok, render = pcall(template.compile, text)
+    if not ok then
+        local err = render
+        err = str_format("%s template compile: %s", typ, err)
+        core.log.error(err)
+        return nil, 503, err
+    end
+
+    out._ctx = ctx
+    out._escape_xml = escape_xml
+    out._escape_json = escape_json
+    local ok, render_out = pcall(render, out)
+    if not ok then
+        local err = str_format("%s template rendering: %s", typ, render_out)
+        core.log.error(err)
+        return nil, 503, err
+    end
+
+    core.log.info(typ, " body transform output=", render_out)
+    return render_out
+end
+
+
+local function set_input_format(conf, typ, ct)
+    if conf[typ].input_format == nil and ct then
+        if ct:find("text/xml") then
+            conf[typ].input_format = "xml"
+        elseif ct:find("application/json") then
+            conf[typ].input_format = "json"
+        end
+    end
+end
+
+
+function _M.rewrite(conf, ctx)
+    if conf.request then
+        conf = core.table.deepcopy(conf)
+        ctx.body_transformer_conf = conf
+        local body = core.request.get_body()
+        set_input_format(conf, "request", ctx.var.http_content_type)
+        local out, status, err = transform(conf, body, "request", ctx)
+        if not out then
+            return status, { message = err }
+        end
+        req_set_body_data(out)
+    end
+end
+
+
+function _M.header_filter(conf, ctx)
+    if conf.response then
+        if not ctx.body_transformer_conf then
+            conf = core.table.deepcopy(conf)
+            ctx.body_transformer_conf = conf
+        end
+        set_input_format(conf, "response", ngx.header.content_type)
+        core.response.clear_header_as_body_modified()
+    end
+end
+
+
+function _M.body_filter(_, ctx)
+    local conf = ctx.body_transformer_conf
+    if conf.response then
```

Comment:
It seems if neither conf.request nor conf.response are given, the conf will be nil here.
We should check if one of them is configured at least in check_schema.